lib/booking_system.rb

```diff
@@ -0,0 +1,118 @@
+module Hotel
+  class BookingSystem
+    attr_accessor :rooms, :reservations, :all_ids, :blocks
+
+    def initialize
+      @rooms = load_rooms
+      @reservations = []
+      @all_ids = []
+      @blocks = []
+    end
+
+    def load_rooms
+      rooms = []
+      i = 1
+
+      20.times do
+        rooms << i
+        i += 1
+      end
+      return rooms
+    end
+
+    def make_reservation(start_date, end_date)
+      reservation = Hotel::Reservation.new(reservation_id: generate_id, room: assign_available_room(start_date, end_date), start_date: start_date, end_date: end_date, price_per_night: 200)
+      @reservations << reservation
+      # binding.pry
+    end
+
+    def make_block(start_date, end_date, number_of_rooms)
+      if number_of_rooms > 5
+        raise ArgumentError, "Cannot reserve more than 5 rooms"
+      end
+
+      id = generate_id
+
+      block = Hotel::Block.new(block_id: id, number_of_rooms: number_of_rooms, start_date: start_date, end_date: end_date, price_per_night: 150)
+      @blocks << block
+
+      number_of_rooms.times do
+        block_reservation = Hotel::BlockReservation.new(block_id: id,  reservation_id: nil, room: assign_available_room(start_date, end_date), start_date: start_date, end_date: end_date, price_per_night: 150)
+        @reservations << block_reservation
+      end
+    end
+
+    def make_block_reservation(block_id)
+      block_reservation = find_empty_block_reservation(block_id)
+      block_reservation.reservation_id = generate_id
+      return block_reservation
+    end
+
+    def assign_available_room(start_date, end_date)
+      booked_rooms = []
+      available_rooms = []
+      @reservations.each do |reservation|
+        if
+          reservation.start_date < end_date && start_date < reservation.end_date
+          booked_rooms << reservation.room
```

Comment:
Here your `BookingSystem` is taking on some extra responsibility: determining whether a reservation overlaps a set of dates. I would argue that that is really the reservation's responsibility. You could write a method called something like `Reservation#overlap?(start_date, end_date)` that takes care of this logic. 

That simplifies this code, and makes it easier to test the logic in isolation. You also repeat this code on line 72, which means that if the requirements change or you discover a bug, you'll have to change it in two places.

Reply:
This also an example of tight coupling, since `BookingSystem` needs to know about how a `Reservation` stores its dates.

Reply:
Why not call `reservations_within_date` here to get the list of `booked_rooms`?